usr.sbin/bhyve/block_if.h

```diff
@@ -75,4 +138,55 @@ int	blockif_delete(struct blockif_ctxt *bc, struct blockif_req *breq);
 int	blockif_cancel(struct blockif_ctxt *bc, struct blockif_req *breq);
 int	blockif_close(struct blockif_ctxt *bc);
 
+/*
+ * Each block device backend registers a set of function pointers that are
+ * used to implement the net backends API.
+ */
+struct block_backend {
+	const char *bb_name;		/* identifier used parse the option string */
+	/*
+	 * Routines used to initialize and cleanup the resources needed
+	 * by a backend. The init and cleanup function are used internally,
+	 * and should not be called by the frontend.
+	 */
+	void	(*bb_init)(void);
+
+	void	(*bb_cleanup)(blockif_ctxt_t *bc);
+
+	blockif_ctxt_t * (*bb_open)(const char *optstr, const char *ident);
+
+        off_t	(*bb_size)(blockif_ctxt_t *bc);
+
+        void	(*bb_chs)(blockif_ctxt_t *bc, uint16_t *c, uint8_t *h, 
+ 		       uint8_t *s);
+
+	int	(*bb_sectsz)(blockif_ctxt_t *bc);
+
+        void	(*bb_psectsz)(blockif_ctxt_t *bc, int *size, int *off);
+
+        int	(*bb_queuesz)(blockif_ctxt_t *bc);
+
+        int	(*bb_is_ro)(blockif_ctxt_t *bc);
+
+        int	(*bb_candelete)(blockif_ctxt_t *bc);
+
+        int	(*bb_read)(blockif_ctxt_t *bc, blockif_req_t *breq);
+
+        int	(*bb_write)(blockif_ctxt_t *bc, blockif_req_t *breq);
+
+        int	(*bb_flush)(blockif_ctxt_t *bc, blockif_req_t *breq);
+
+        int	(*bb_delete)(blockif_ctxt_t *bc, blockif_req_t *breq);
+
+        int	(*bb_cancel)(blockif_ctxt_t *bc, blockif_req_t *breq);
+
+        int	(*bb_close)(blockif_ctxt_t *bc);
+
+	/* Room for backend-specific data. */
+	void *bb_opaque;
+};
+
+/* Cheating for legacy code were we need the blocklocal_backend block */
+block_backend_t blocklocal_backend;
```

Comment:
I can't understand this comment.

Reply:
I'll try to refrase this, but it is a poor attempt not to create a `block_local.h` file with only one thing in it: the address of the jump table.
And that is needed in `block_if.c`, to be able to use it in case of a legacy PCI-slot specification without `file:`
The other way around would be to prefix legacy with `file:`, and run it through the regular backend-search loop. But detecting legacy by scanning the string is rather complex or error prone.

Reply:
Actually, the "way around" looks fine to me. Something like this:
```
backend_name = "file";
optrest = strchr(optstr, ':');
if (optrest != NULL) {
  *optrest = '\0';
  backend_name = optstr;
  optstr = ++optrest;
}
/* search for backend using backend_name */
```